SIPS/sip-14.md

```diff
@@ -0,0 +1,86 @@
+---
+sip: 14
+title: Dynamic Permissions
+status: Draft
+discussions-to: 
+author: Frederik Bolding (@frederikbolding)
+created: 2023-10-19
+---
+
+## Abstract
+
+This SIP proposes changes to the Snap manifest and new RPC methods that would allow Snap developers to request additional permissions dynamically at runtime. This proposal outlines some of the details around this feature.
+
+## Motivation
+
+Snaps currently have to request all permissions that they plan to use at install-time. This becomes a problem when a Snap wants to use many permissions as the installation experience suffers and the user has to either accept all permissions requested or deny the installation. This proposal provides an improvement to the experience by letting Snaps request permissions at run-time as long as those permissions are statically defined in the manifest at build-time.

```

Comment:
Yes, that is true. But `eth_accounts` is kind of an edge-case right now. We should probably require it being put into `dynamicPermissions`

Reply:
That would be a new permission? Right now we just require that they request `endowment:ethereum-provider`, so it's like a dynamic permission on top of a static one.

Reply:
`eth_accounts` is an edge-case because it is technically a permission that is granted when using certain RPC methods with `endowment:ethereum-provider`. It is not inherently granted just by having the endowment.

We could choose to enforce that you need to have it as part of `dynamicPermissions` to grant it via `endowment:ethereum-provider`. We could also choose to keep it as the edge-case that it is.

Reply:
I do wonder if we have created a strange design pattern with the current approach. What if a Snap just wants `eth_accounts` and is requesting `endowment:ethereum-provider` just for that, and not using anything else that the endowment offers? Maybe it should be a separate endowment that must be defined in `dynamicPermissions` and is never in `initialPermissions`. Assuming this change would be in manifest version 0.2 and would not be backwards compatible.